### ISSUE TITLE
Yolov8 segmentation

### ISSUE THREAD

**Issue by Manueljohnson063** [ultralytics/ultralytics#5180]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Omitting the bbox from the segmentation results  for improving the performance .

### Additional

Hi,
First of all thanks for this amazing repo!

1.I am currently working with **yolov8n segmentation** .When i tried to display the results i have noticed that the segmentation have Bounding boxes.While annotating i have not included the bbox .Then why it is coming in output , My question is how to get  the segmentation alone(Not asking for commenting the bbox ,**_ but In the architecture level_**).
2.I am runnning  yolov8n-seg in Jetson Xavier because of that i need to optimise the model to the maximum ,does this bounding box detection in inference takes much time ?If so how to omit that from architecture?So my model will generate only the  segmentation results.
Thanks in Advance :)@glenn-jocher

**Comment by RizwanMunawar** [comment]
@Manueljohnson063 Object segmentation and pose estimation have been integrated into the object detection wrapper. Consequently, the bounding box is included by default, and the only way to remove it is by using the `boxes=False` argument.

Hope this helps! Let me know if you have any additional questions.
Thanks

**Comment by Manueljohnson063** [comment]
Thanks for your quick replay.
The reason why i asked is that if i can remove the object detection process in segmentation ,though i  can reduce the inference time.
I am not using the ultralytcs inference code ,Because i need to deploy the model in the Jetson device.
Thanks in advance

**Comment by glenn-jocher** [comment]
Hello @Manueljohnson063,

Thank you for your explanation.

Firstly, I'd like to clarify that in our YOLOv8 architecture the segmentation process is not separate from the detection process. They are part of the same pipeline where detection (drawing bounding boxes) and segmentation (identifying each pixel's class) are tightly intertwined. This means removing the detection process could potentially lead to incorrect or unrefined segmentation results. 

If you are looking to reduce the inference time, you might consider optimizing other parts of the pipeline instead. Possible options include simplifying your model architecture, reducing input resolution, or implementing hardware-specific optimizations for Jetson devices. 

In the context of deploying the model in a Jetson device using C++, while I don't have a specific example to guide you, generally you would go through similar steps: load the model, prepare your image data, and run inference. Remember to set the `boxes=False` argument in your code logic to prevent bounding boxes from being drawn, if that's what you prefer.

Remember that while removal of bounding boxes might save some processing time, in the holistic view of the entire processing chain this time might be insignificant compared to the time taken for segmentation.

I hope this clarifies it for you! Don't hesitate to reach out if you have any more questions. Happy coding!

**Comment by Manueljohnson063** [comment]
When i had netron on the onnx model(yolov8n-seg.onnx) i am able to notice that there are two outputs from the network output0(1x32x160x160) and output1(1x116x8400) .
Can you please give me an idea about the dimensions and the output .

**Comment by wenlanc** [comment]
> 1. Width x Height

I always get this result '1x32x160x160' from any image.
Here 'Number of Classes' means the size of detected classes?
If so, how to get class id, I have got only 7 classes.
Please help me.

**Comment by glenn-jocher** [comment]
@wenlanc hey there! 🙂

The `1x32x160x160` you're seeing represents the segmentation output format where `32` represents the predicted classes layer. If you're working with only 7 classes, it sounds like the model might be configured for a broader class set than you need.

In the segmentation output, each pixel in the `160x160` grid is categorized into one of those 32 classes. To focus on your 7 classes, you'd need to map these output class indices to your class IDs based on the dataset you trained on.

Here's a little tip: if you're diving directly into the raw output, the class ID would correspond to the channel with the highest value for a given pixel. So, for each pixel in your `160x160` grid, find the max value across the `32` channels - that index is the predicted class for the pixel.

Hope that helps! If you need more detailed guidance or a code snippet on handling these outputs, feel free to ask. Happy coding! 🚀

**Comment by wenlanc** [comment]
Hi @glenn-jocher 
Thanks your kindly reply.
Can you help me with code snippet on handling these outputs?

**Comment by glenn-jocher** [comment]
Hi @wenlanc, glad to help! 😊 To handle the segmentation outputs (`1x32x160x160`) for your 7 classes, you can simply map the index of the max value in the 32 channels to your class IDs. Here's a quick example using PyTorch:

```python
import torch

# Sample output tensor simulating '1x32x160x160' output
output = torch.rand(1, 32, 160, 160)  # Random tensor for demonstration

# Get predicted class per pixel (find max index across 32 channels)
predicted_classes = torch.argmax(output, dim=1)

# Now, predicted_classes is a 1x160x160 tensor with class IDs as values
```

This `predicted_classes` tensor will have the class index of the highest value for each pixel. If your classes are a subset, just map these indices back to your specific class IDs.

Hope this helps! Let me know if you need further assistance. 🚀

**Comment by wenlanc** [comment]
Thanks you.
but I got failed result image yet.
My code like this
`CLASSES = {
    0: 'AAA',
    1: 'BBB',
    2: 'CCC',
    3: 'DDD',
    4: 'EEE',
    5: 'FFF',
    6: 'GGG'
}
colors = np.random.uniform(0, 255, size=(len(CLASSES), 3))

image = cv.imread('sample.jpg')
img_height, img_width = image.shape[:2]
input_Width, input_Height = 640, 640
image = cv.resize(image, (input_Width, input_Height))

model = cv.dnn.readNetFromONNX("best.onnx")
blob = cv.dnn.blobFromImage(image=image, size=(input_Width, input_Height), mean=(0, 0, 0), swapRB=True, crop=False)
model.setInput(blob)
output0 = model.forward(outputName='output0')
output1 = model.forward(outputName='output1')

output0 = output0[0].transpose()
output1 = output1[0]

boxes = output0[:, 0:11]
masks = output0[:, 11:]
print("Boxes:", boxes.shape, "Masks:", masks.shape)

output1 = output1.reshape(32, 160*160)
print(masks.shape, output1.shape)

# Join matrices by matrix multiplication
masks = masks @ output1
print(masks.shape)

# Connect masks to boxes: move all columns from masks matrix to the right of boxes matrix
boxes = np.hstack([boxes, masks])
print(boxes.shape)

# Parse "boxes" to get list of bounding boxes of detected objects,
# their segmentation masks and bounding polygons

def intersection(box1, box2):
    box1_x1, box1_y1, box1_x2, box1_y2 = box1[:4]
    box2_x1, box2_y1, box2_x2, box2_y2 = box2[:4]
    x1 = max(box1_x1, box2_x1)
    y1 = max(box1_y1, box2_y1)
    x2 = min(box1_x2, box2_x2)
    y2 = min(box1_y2, box2_y2)
    return (x2-x1)*(y2-y1)


def union(box1, box2):
    box1_x1, box1_y1, box1_x2, box1_y2 = box1[:4]
    box2_x1, box2_y1, box2_x2, box2_y2 = box2[:4]
    box1_area = (box1_x2-box1_x1)*(box1_y2-box1_y1)
    box2_area = (box2_x2-box2_x1)*(box2_y2-box2_y1)
    return box1_area + box2_area - intersection(box1, box2)


def iou(box1, box2):
    return intersection(box1, box2)/union(box1, box2)


def sigmoid(z):
    return 1/(1 + np.exp(-z))


# parse segmentation mask
def get_mask(row, box, img_width, img_height):
    # convert mask to image (matrix of pixels)
    mask = row.reshape(160, 160)
    mask = sigmoid(mask)
    mask = (mask > 0.5).astype("uint8")*255
    # crop the object defined by "box" from mask
    x1, y1, x2, y2 = box
    mask_x1 = round(x1/img_width*160)
    mask_y1 = round(y1/img_height*160)
    mask_x2 = round(x2/img_width*160)
    mask_y2 = round(y2/img_height*160)
    mask = mask[mask_y1:mask_y2, mask_x1:mask_x2]
    # resize the cropped mask to the size of object
    img_mask = Image.fromarray(mask, "L")
    img_mask = img_mask.resize((round(x2-x1), round(y2-y1)))
    mask = np.array(img_mask)
    return mask


# calculate bounding polygon from mask
def get_polygon(mask):
    contours = cv.findContours(mask, cv.RETR_LIST, cv.CHAIN_APPROX_SIMPLE)
    polygon = [[contour[0][0], contour[0][1]] for contour in contours[0][0]]
    return polygon


# parse and filter all boxes
objects = []
for row in boxes:
    xc, yc, w, h = row[:4]
    x1 = (xc-w/2)/640*img_width
    y1 = (yc-h/2)/640*img_height
    x2 = (xc+w/2)/640*img_width
    y2 = (yc+h/2)/640*img_height
    prob = row[4:11].max()
    if prob < 0.5:
        continue
    class_id = row[4:11].argmax()
    label = yolo_classes[class_id]
    mask = get_mask(row[11:25611], (x1, y1, x2, y2), img_width, img_height)
    if mask.max() == 0:
        continue
    polygon = get_polygon(mask)
    objects.append([x1, y1, x2, y2, class_id, label, prob, mask, polygon])


def draw_bounding_box(img, class_id, confidence, x, y, x_plus_w, y_plus_h):
    """
    Draws bounding boxes on the input image based on the provided arguments.

    Args:
        img (numpy.ndarray): The input image to draw the bounding box on.
        class_id (int): Class ID of the detected object.
        confidence (float): Confidence score of the detected object.
        x (int): X-coordinate of the top-left corner of the bounding box.
        y (int): Y-coordinate of the top-left corner of the bounding box.
        x_plus_w (int): X-coordinate of the bottom-right corner of the bounding box.
        y_plus_h (int): Y-coordinate of the bottom-right corner of the bounding box.
    """
    label = f"{CLASSES[class_id]} ({confidence:.2f})"
    color = colors[class_id]
    cv.rectangle(img, (int(x), int(y)), (int(x_plus_w), int(y_plus_h)), color, 2)
    cv.putText(img, label, (int(x - 10), int(y - 10)), cv.FONT_HERSHEY_SIMPLEX, 0.5, color, 2)

# apply non-maximum suppression
objects.sort(key=lambda x: x[5], reverse=True)
result = []
while len(objects) > 0:
    result.append(objects[0])
    objects = [object for object in objects if iou(object, objects[0]) < 0.7]

# display bounding polygons of detected objects on the image
for object in result:
    [x1, y1, x2, y2, class_id, label, prob, mask, polygon] = object
    draw_bounding_box(image, class_id, prob, x1, y1, x2, y2)

cv.imshow("Result", image)
cv.waitKey(0)
`

**Comment by wenlanc** [comment]
Hi @glenn-jocher 
When I use onnxruntime and cv2.cnn. readNetFromONNX, the outputs structure is same, but the values are different.
output0: 1*43*8400
output1: 1*32*160*160
here 4(xc, yc, w, h) + 7(Number of classes) + 32(masks), right?
How to handle these outputs?
Thanks.

**Comment by glenn-jocher** [comment]
@wenlanc hi there! 👋 

Your observation is correct! The output structure you're seeing is typical for a YOLOv8 model with segmentation capabilities:

- `output0` (1x43x8400) contains information for object detection, including coordinates (`xc, yc, w, h`), class probabilities (for 7 classes in your case), and additional data.
- `output1` (1x32x160x160) is for segmentation masks, with the `32x160x160` format representing different mask channels for each class.

To handle these outputs, you'd typically do the following:

1. **For Detection (`output0`):** Interpret each of the 8400 vectors to decode bounding boxes and class predictions. Here's a simplified approach:
```python
detections = output0.reshape(-1, 43)  # Reshape for easier parsing
for detection in detections:
    xc, yc, w, h, *class_probs = detection[:11]  # Split detection vector
    # Convert from center coordinates to bounding box corners here
    # Use class_probs to determine the object class
```

2. **For Segmentation (`output1`):** You can apply a threshold to decide which pixels belong to which class. This could look something like:
```python
segmentation = torch.argmax(output1, dim=1)  # Assuming output1 is a tensor
```
This would give you the most likely class for each pixel in the 160x160 grid.

🌟 Remember, handling the outputs involves various post-processing steps like Non-Max Suppression (NMS) for detection outputs, and scaling up the segmentation masks to your original image size.

Hope this helps! Let me know if you have further questions. 🚀

**Comment by AliBerkKaraarslanDO** [comment]
Hello @glenn-jocher!
I have few questions. If I am not mistaken, yolov8n-seg has 80 classes. But its' output has 32 layers. I wonder how do map this 32 layers to 80 classes? Also I have a yolov8 segmentation model with 3 classes. I want to represent the segmentation mask using output layers but I am not sure how to map this 32 layers to 3 classes. Can you help?
Thanks.

**Comment by glenn-jocher** [comment]
Hello!

Great questions! In YOLOv8n-seg, the 32 layers in the output don't directly correspond to the 80 classes for detection. Instead, these layers represent different features learned by the network, which are used to construct the segmentation masks.

For mapping these layers to your specific 3 classes in a segmentation task, you typically use a softmax or sigmoid function across these layers to determine the class probabilities for each pixel. Here’s a simple way to do this using PyTorch:

```python
import torch.nn.functional as F

# Assuming 'output' is your model's segmentation output with shape [1, 32, H, W]
output_probs = F.softmax(output, dim=1)  # Apply softmax on the channel dimension
class_masks = output_probs.argmax(dim=1)  # Get the class with the highest probability for each pixel
```

This `class_masks` tensor will now hold the class IDs (0, 1, or 2) for each pixel in your image.

Hope this helps! Let me know if you need further clarification. 😊

**Comment by AliBerkKaraarslanDO** [comment]
@glenn-jocher Thank you for your answer!
I am using yolov8n-seg model and I did the softmax process like you mentioned earlier. My `class_masks ` tensor's dimensions are 1x160x160 but the values are between 0-31. How can I understand the class IDs from these values?  Do I need to use output layer with the size of 1x116x8400? Thanks a lot.

**Comment by glenn-jocher** [comment]
Hi there!

It sounds like you're on the right track! The values in your `class_masks` tensor ranging from 0-31 correspond to the indices of the highest probability class at each pixel, based on the 32 output channels of your segmentation model. These indices represent different learned features or potential class segments, not direct class IDs from your dataset.

If your model is intended to segment 3 classes, you should ensure that the final layer of your model maps these 32 channels to your 3 specific classes. This might involve adjusting the model architecture or the training process to align the output with your class IDs.

You don't need to use the detection output layer (1x116x8400) for segmentation tasks unless you're also interested in bounding box predictions alongside segmentation masks.

If you need further adjustments or clarification on how to modify your model for correct class mapping, feel free to ask. Happy coding! 😊

**Comment by AliBerkKaraarslanDO** [comment]
@glenn-jocher Thank you for our clarification!
But still I am not sure how to do the mapping. Can you give me more details?
Thank you for your patience.
